Doc/howto/a-conceptual-overview-of-asyncio.rst

```diff
@@ -0,0 +1,736 @@
+
+        while True:
+            try:
+                # For reference, the first argument to recv() is the maximum number
+                # of bytes to attempt to read. Setting it to 4096 means we could get 2
+                # bytes or 4 bytes, or even 4091 bytes, but not 4097 bytes back.
+                # However, if there are no bytes available to be read, this recv()
+                # will raise a BlockingIOError since the socket was set to
+                # non-blocking mode.
+                response = client.recv(4096)
+                break
+            except BlockingIOError:
+                await YieldToEventLoop()
+        return response
+
+
+And this is the portion of code responsible for asynchronously computing
+the uniform sums. It's designed to allow for working through the sum a portion at a time.
+The ``time_allotment`` argument to the coroutine function decides how long the sum
+function will iterate, in other words, synchronously hog control, before ceding
+back to the event-loop.
+
+.. code-block:: python
+
+    async def uniform_sum(n_samples: int, time_allotment: float) -> float:
+
+        start_time = time.time()
+
+        total = 0.0
+        for _ in range(n_samples):
+            total += random.random()
+
+            time_elapsed = time.time() - start_time
+            if time_elapsed > time_allotment:
+                await YieldToEventLoop()
+                start_time = time.time()
+
+        return total
+
+The above snippet was simplified a bit. Reading ``time.time()`` and evaluating an if-condition
+on every iteration for many, many iterations (in this case roughly a hundred million) more than
+eats up the runtime savings associated with the asynchronous approach. The actual
+implementation involves chunking the iteration, so you only perform the check
+every few million iterations. With that change, the asynchronous approach wins in a landslide.
+This is important to keep in mind. Too much checking or constantly jumping between tasks can
+ultimately cause more harm than good!
+
+The server, async & serial programs are available in full here:
+https://github.com/anordin95/a-conceptual-overview-of-asyncio/tree/main/barebones-network-io-example.
+
+.. _which_concurrency_do_I_want:
+
+------------------------------
+Which concurrency do I want
+------------------------------
+
+===========================
+multiprocessing
+===========================
+
+For any computationally bound work in Python, you likely want to use multiprocessing.
+Otherwise, the Global Interpreter Lock (GIL) will generally get in your way! For those who don't
+know, the GIL is a lock which ensures only one Python instruction is executed at a time.
+Of course, since processes are generally entirely independent from one another, the GIL in one
+process won't be impeded by the GIL in another process. Granted, I believe there are ways to
+also get around the GIL in a single process by leveraging C extensions.
```

Comment:
This should cover subinterpreters. However, in the meantime I would probably remove this section -- this PR is alredy very large (750 lines).

Reply:
I hate to vaccillate, but I'm having second thoughts about whacking the "Which concurrency do I want" in the interest of size. I think it offers accessible, simple, high-level context for guiding folks towards the right tool. And the section is a mere 330 words of the article's over 4,000 words. 

And, no I have not seen Eric's draft general concurrency HOWTO, but I'd be curious to look at it! It probably makes sense to link to it from this section once it's done.

And, I've added a brief reference to using subinterpreters.

Reply:
> And the section is a mere 330 words of the article's over 4,000 words.

To be slightly blunt, the risk you face with this PR is that it will go unreviewed due to lack of time/resource. Any committer will generally look askance at such a large diff, even if it is 'only' documentation. My advice here is more general: you will have a better chance of making progress with a series of smaller changes. The asyncio tutorial we discussed earlier didn't languish due to content reasons, but for process -- a request to remove some sections here isn't binning those words, just delaying them to the next PR.

A

Reply:
Ahhh, I see. Fair enough! I went ahead and took it a step further by cutting the two example sections & the "which concurrency do I want" section. It's now roughly half the original size.